File: core/src/chunker.ts
/**
 * Document chunker - splits documents into analyzable pieces
 */

import type { ParsedDocument, Chunk } from './types.js';

export interface ChunkOptions {
  /** Preferred chunk strategy */
  strategy: 'chapters' | 'size' | 'hybrid';
  /** Target chunk size in characters (for 'size' and 'hybrid' strategies) */
  targetSize?: number;
  /** Maximum chunk size (hard limit) */
  maxSize?: number;
  /** Minimum chunk size (avoid tiny chunks) */
  minSize?: number;
}

const DEFAULT_OPTIONS: Required<ChunkOptions> = {
  strategy: 'hybrid',
  targetSize: 8000,   // ~2000 tokens, good for detailed analysis
  maxSize: 16000,     // ~4000 tokens max
  minSize: 1000,      // ~250 tokens min
};

/**
 * Split a parsed document into chunks for analysis
 */
export function chunkDocument(
  doc: ParsedDocument,
  options: Partial<ChunkOptions> = {}
): Chunk[] {
  const opts = { ...DEFAULT_OPTIONS, ...options };

  switch (opts.strategy) {
    case 'chapters':
      return chunkByChapters(doc, opts);
    case 'size':
      return chunkBySize(doc, opts);
    case 'hybrid':
    default:
      return chunkHybrid(doc, opts);
  }
}

/**
 * Chunk by detected chapter/section headings
 */
function chunkByChapters(
  doc: ParsedDocument,
  opts: Required<ChunkOptions>
): Chunk[] {
  const chunks: Chunk[] = [];

  if (doc.headings.length === 0) {
    // No headings found, fall back to size-based chunking
    return chunkBySize(doc, opts);
  }

  // Find top-level headings (lowest level number = highest importance)
  const minLevel = Math.min(...doc.headings.map(h => h.level));
  const chapterHeadings = doc.headings.filter(h => h.level <= minLevel + 1);

  for (let i = 0; i < chapterHeadings.length; i++) {
    const heading = chapterHeadings[i];
    const nextHeading = chapterHeadings[i + 1];

    const startIndex = heading.paragraphIndex;
    const endIndex = nextHeading
      ? nextHeading.paragraphIndex
      : doc.paragraphs.length;

    const content = doc.paragraphs.slice(startIndex, endIndex).join('\n\n');

    chunks.push({
      id: `chunk-${i + 1}`,
      title: heading.text,
      content,
      startIndex,
      endIndex,
    });
  }

  // Handle any content before the first heading
  if (chapterHeadings[0]?.paragraphIndex > 0) {
    const preambleContent = doc.paragraphs
      .slice(0, chapterHeadings[0].paragraphIndex)
      .join('\n\n');

    if (preambleContent.trim().length > opts.minSize) {
      chunks.unshift({
        id: 'chunk-0',
        title: 'Preamble',
        content: preambleContent,
        startIndex: 0,
        endIndex: chapterHeadings[0].paragraphIndex,
      });
    }
  }

  return chunks;
}

/**
 * Chunk by target size, breaking at paragraph boundaries
 */
function chunkBySize(
  doc: ParsedDocument,
  opts: Required<ChunkOptions>
): Chunk[] {
  const chunks: Chunk[] = [];
  let currentChunk: string[] = [];
  let currentSize = 0;
  let startIndex = 0;
  let chunkNum = 1;

  for (let i = 0; i < doc.paragraphs.length; i++) {
    const para = doc.paragraphs[i];
    const paraSize = para.length;

    // If adding this paragraph would exceed max, start a new chunk
    if (currentSize + paraSize > opts.maxSize && currentChunk.length > 0) {
      chunks.push({
        id: `chunk-${chunkNum}`,
        title: null,
        content: currentChunk.join('\n\n'),
        startIndex,
        endIndex: i,
      });
      chunkNum++;
      currentChunk = [];
      currentSize = 0;
      startIndex = i;
    }

    currentChunk.push(para);
    currentSize += paraSize;

    // If we've reached target size, consider ending the chunk
    if (currentSize >= opts.targetSize) {
      // Look ahead for a good break point (paragraph that looks like section end)
      const isGoodBreak = isParagraphBreak(para) ||
        (i + 1 < doc.paragraphs.length && isParagraphBreak(doc.paragraphs[i + 1]));

      if (isGoodBreak || currentSize >= opts.maxSize) {
        chunks.push({
          id: `chunk-${chunkNum}`,
          title: null,
          content: currentChunk.join('\n\n'),
          startIndex,
          endIndex: i + 1,
        });
        chunkNum++;
        currentChunk = [];
        currentSize = 0;
        startIndex = i + 1;
      }
    }
  }

  // Don't forget the last chunk
  if (currentChunk.length > 0) {
    chunks.push({
      id: `chunk-${chunkNum}`,
      title: null,
      content: currentChunk.join('\n\n'),
      startIndex,
      endIndex: doc.paragraphs.length,
    });
  }

  return chunks;
}

/**
 * Hybrid: use chapters when available, subdivide large chapters by size
 */
function chunkHybrid(
  doc: ParsedDocument,
  opts: Required<ChunkOptions>
): Chunk[] {
  const chapterChunks = chunkByChapters(doc, opts);
  const result: Chunk[] = [];

  for (const chunk of chapterChunks) {
    if (chunk.content.length <= opts.maxSize) {
      result.push(chunk);
    } else {
      // Subdivide large chapters
      const subDoc: ParsedDocument = {
        title: chunk.title || 'Untitled',
        fullText: chunk.content,
        paragraphs: chunk.content.split('\n\n').filter(p => p.trim()),
        headings: [],
      };

      const subChunks = chunkBySize(subDoc, opts);

      for (let i = 0; i < subChunks.length; i++) {
        result.push({
          ...subChunks[i],
          id: `${chunk.id}-${i + 1}`,
          title: chunk.title ? `${chunk.title} (Part ${i + 1})` : null,
          // Adjust indices to be relative to original document
          startIndex: chunk.startIndex + subChunks[i].startIndex,
          endIndex: chunk.startIndex + subChunks[i].endIndex,
        });
      }
    }
  }

  return result;
}

/**
 * Heuristic: does this paragraph look like a section break?
 */
function isParagraphBreak(para: string): boolean {
  const trimmed = para.trim();

  // Scene break markers
  if (/^[*#-]{3,}$/.test(trimmed)) return true;
  if (trimmed === '***' || trimmed === '---' || trimmed === '###') return true;

  // Very short paragraph (often indicates break)
  if (trimmed.length < 20) return true;

  // Ends with strong punctuation
  if (/[.!?]["']?$/.test(trimmed)) return true;

  return false;
}

/**
 * Utility: estimate token count (rough approximation)
 */
export function estimateTokens(text: string): number {
  // Rough estimate: ~4 characters per token for English prose
  return Math.ceil(text.length / 4);
}

File: critic/chunker.py
"""
Document chunker - splits documents into analyzable sections.
"""

from typing import Literal, Optional
from dataclasses import dataclass

from critic.types import ParsedDocument, Chunk, Heading


@dataclass
class ChunkOptions:
    """Options for chunking."""
    strategy: Literal["chapters", "size", "hybrid"] = "hybrid"
    target_size: int = 8000  # characters
    min_size: int = 2000
    max_size: int = 15000


def estimate_tokens(text: str) -> int:
    """
    Estimate token count for text.
    Rough approximation: ~4 characters per token for English.
    """
    return len(text) // 4


def chunk_document(
    doc: ParsedDocument,
    options: Optional[ChunkOptions] = None,
) -> list[Chunk]:
    """
    Chunk a document into sections for analysis.

    Args:
        doc: Parsed document
        options: Chunking options

    Returns:
        List of chunks
    """
    opts = options or ChunkOptions()

    if opts.strategy == "chapters":
        return _chunk_by_chapters(doc, opts)
    elif opts.strategy == "size":
        return _chunk_by_size(doc, opts)
    else:  # hybrid
        return _chunk_hybrid(doc, opts)


def _is_chapter_heading(heading: Heading) -> bool:
    """Check if a heading is a chapter boundary."""
    # Level 1-2 headings are always chapter boundaries
    if heading.level <= 2:
        return True
    # Level 3 headings that look like chapter titles
    text_lower = heading.text.lower()
    if heading.level == 3 and (
        text_lower.startswith('chapter') or
        text_lower.startswith('epilogue') or
        text_lower.startswith('prologue') or
        text_lower.startswith('part ')
    ):
        return True
    return False


def _chunk_by_chapters(doc: ParsedDocument, opts: ChunkOptions) -> list[Chunk]:
    """Chunk by chapter headings."""
    if not doc.headings:
        # No headings, fall back to size-based
        return _chunk_by_size(doc, opts)

    # Find chapter-level headings
    chapter_headings = [h for h in doc.headings if _is_chapter_heading(h)]

    # Skip title/subtitle headings at the start (paragraph 0)
    chapter_headings = [h for h in chapter_headings if h.paragraph_index > 0]

    if not chapter_headings:
        return _chunk_by_size(doc, opts)

    chunks: list[Chunk] = []

    for i, heading in enumerate(chapter_headings):
        # For first chunk, include any content before the first chapter heading
        start_para = 0 if i == 0 else heading.paragraph_index
        end_para = (
            chapter_headings[i + 1].paragraph_index
            if i + 1 < len(chapter_headings)
            else len(doc.paragraphs)
        )

        content = '\n\n'.join(doc.paragraphs[start_para:end_para])

        chunks.append(Chunk(
            id=f"chunk-{i + 1}",
            title=heading.text,
            content=content,
            start_paragraph=start_para,
            end_paragraph=end_para,
        ))

    return chunks


def _chunk_by_size(doc: ParsedDocument, opts: ChunkOptions) -> list[Chunk]:
    """Chunk by target size."""
    chunks: list[Chunk] = []
    current_content: list[str] = []
    current_size = 0
    start_para = 0
    chunk_num = 1

    for i, para in enumerate(doc.paragraphs):
        para_size = len(para)

        # Check if adding this paragraph would exceed max size
        if current_size + para_size > opts.max_size and current_content:
            # Save current chunk
            chunks.append(Chunk(
                id=f"chunk-{chunk_num}",
                title=None,
                content='\n\n'.join(current_content),
                start_paragraph=start_para,
                end_paragraph=i,
            ))
            chunk_num += 1
            current_content = []
            current_size = 0
            start_para = i

        current_content.append(para)
        current_size += para_size

        # Check if we've reached target size
        if current_size >= opts.target_size:
            chunks.append(Chunk(
                id=f"chunk-{chunk_num}",
                title=None,
                content='\n\n'.join(current_content),
                start_paragraph=start_para,
                end_paragraph=i + 1,
            ))
            chunk_num += 1
            current_content = []
            current_size = 0
            start_para = i + 1

    # Don't forget remaining content
    if current_content:
        chunks.append(Chunk(
            id=f"chunk-{chunk_num}",
            title=None,
            content='\n\n'.join(current_content),
            start_paragraph=start_para,
            end_paragraph=len(doc.paragraphs),
        ))

    return chunks


def _chunk_hybrid(doc: ParsedDocument, opts: ChunkOptions) -> list[Chunk]:
    """
    Hybrid chunking: respect chapter boundaries but split large chapters.
    """
    # First, get chapter-based chunks
    chapter_chunks = _chunk_by_chapters(doc, opts)

    if not chapter_chunks:
        return _chunk_by_size(doc, opts)

    # Split any chunks that are too large
    final_chunks: list[Chunk] = []
    chunk_num = 1

    for chapter in chapter_chunks:
        if len(chapter.content) <= opts.max_size:
            final_chunks.append(Chunk(
                id=f"chunk-{chunk_num}",
                title=chapter.title,
                content=chapter.content,
                start_paragraph=chapter.start_paragraph,
                end_paragraph=chapter.end_paragraph,
            ))
            chunk_num += 1
        else:
            # Split this chapter
            sub_chunks = _split_large_chunk(chapter, opts, chunk_num)
            final_chunks.extend(sub_chunks)
            chunk_num += len(sub_chunks)

    return final_chunks


def _split_large_chunk(
    chunk: Chunk,
    opts: ChunkOptions,
    start_num: int,
) -> list[Chunk]:
    """Split a large chunk into smaller pieces."""
    paragraphs = chunk.content.split('\n\n')
    sub_chunks: list[Chunk] = []
    current_content: list[str] = []
    current_size = 0
    sub_num = 1

    for para in paragraphs:
        para_size = len(para)

        if current_size + para_size > opts.target_size and current_content:
            sub_chunks.append(Chunk(
                id=f"chunk-{start_num + sub_num - 1}",
                title=f"{chunk.title} (Part {sub_num})" if chunk.title else None,
                content='\n\n'.join(current_content),
                start_paragraph=chunk.start_paragraph,
                end_paragraph=chunk.end_paragraph,
            ))
            sub_num += 1
            current_content = []
            current_size = 0

        current_content.append(para)
        current_size += para_size

    if current_content:
        sub_chunks.append(Chunk(
            id=f"chunk-{start_num + sub_num - 1}",
            title=f"{chunk.title} (Part {sub_num})" if chunk.title else None,
            content='\n\n'.join(current_content),
            start_paragraph=chunk.start_paragraph,
            end_paragraph=chunk.end_paragraph,
        ))

    return sub_chunks
